Lib/ctypes/__init__.py

```
@@ -266,7 +266,71 @@ class c_void_p(_SimpleCData):
 class c_bool(_SimpleCData):
     _type_ = "?"
 
-from _ctypes import POINTER, pointer, _pointer_type_cache
+def POINTER(cls):
+    """Create and return a new ctypes pointer type.
+
+    Pointer types are cached and reused internally,
+    so calling this function repeatedly is cheap.
+    """
+    if cls is None:
+        return c_void_p
+    try:
+        pt = cls.__pointer_type__
+        if pt is not None:
+            return pt
+    except AttributeError:
+        pass
+    if isinstance(cls, str):
+        # handle old-style incomplete types
+        # in this case pointer type is not cached and calling this function
+        # repeatedly will give different result
+        return type(f'LP_{cls}', (_Pointer,), {})
+
+    # create pointer type and set __pointer_type__ for cls
+    return type(f'LP_{cls.__name__}', (_Pointer,), {'_type_': cls})
+
+def pointer(obj):
+    """Create a new pointer instance, pointing to 'obj'.
+
+    The returned object is of the type POINTER(type(obj)). Note that if you
+    just want to pass a pointer to an object to a foreign function call, you
+    should use byref(obj) which is much faster.
+    """
+    typ = POINTER(type(obj))
+    return typ(obj)
+
+class PointerTypeCache:
+    def __setitem__(self, cls, pointer_type):
+        import warnings
+        warnings._deprecated("ctypes._pointer_type_cache", remove=(3, 19))
+        try:
+            cls.__pointer_type__ = pointer_type
+        except AttributeError:
+            pass
```

Strongly -1. This just denies the purpose of `AttributeError`.

---

I'm unsure. Previous implementation allows silently rewrite entry in the `_pointer_type_cache`. If we raise exception here it may break some code (maybe it is worth to do). If we silently swallow exception, then it may add some strange heisenbugs.

---

Then, is there any reason why the C setter is not the right place to supress the error?